File: apps/desktop/packages/shared/src/validation/agent_turns.ts
import type {
  AgentAssistantRow,
  AgentPendingStateRow,
  AgentToolCallRow,
  AgentTurnRow,
  AgentTurnsPageQuery,
  AgentTurnsPageResult,
  SessionId,
} from "../contracts.js";
import { PROTOCOL_JSON_SCHEMAS } from "../../generated/runtime.js";
import {
  ajv,
  formatProtocolValidationErrors,
  parseNonEmptyProtocolString,
  parseNullableBoundaryValue,
  parseProtocolBigInt,
  parseSchema,
  parseStringField,
  ProtocolValidationError,
} from "./core.js";
import { parseActivityCursor, parseDaemonEventCursor } from "./cursors.js";
import { parseSessionId } from "./identity.js";

const validateAgentTurnsPageQuery = ajv.compile(PROTOCOL_JSON_SCHEMAS.AgentTurnsPageQuery);
const validateAgentTurnsPageResult = ajv.compile(PROTOCOL_JSON_SCHEMAS.AgentTurnsPageResult);

export function parseAgentTurnsPageQuery(value: unknown): AgentTurnsPageQuery {
  const record = parseSchema<{
    limit: number;
    before?: unknown;
  }>("AgentTurnsPageQuery", validateAgentTurnsPageQuery, value);
  return {
    limit: record.limit,
    before: parseNullableBoundaryValue(record.before, (before) =>
      parseActivityCursor(before, "AgentTurnsPageQuery.before"),
    ),
  };
}

export function parseAgentTurnsPageResult(value: unknown): AgentTurnsPageResult {
  if (validateAgentTurnsPageResult(value)) {
    const record = value as {
      items?: unknown[];
      nextBefore?: unknown;
      latestCursor?: unknown;
    };
    return {
      items: (record.items ?? []).map((item) => parseAgentTurnRow(item)),
      nextBefore: parseNullableBoundaryValue(record.nextBefore, (nextBefore) =>
        parseActivityCursor(nextBefore, "AgentTurnsPageResult.nextBefore"),
      ),
      latestCursor: parseNullableBoundaryValue(record.latestCursor, (latestCursor) =>
        parseDaemonEventCursor(latestCursor, "AgentTurnsPageResult.latestCursor"),
      ),
    };
  }

  throw new ProtocolValidationError(
    formatProtocolValidationErrors("AgentTurnsPageResult", validateAgentTurnsPageResult.errors),
  );
}

function parseAgentTurnRow(value: unknown): AgentTurnRow {
  if (typeof value !== "object" || value === null || !("kind" in value)) {
    throw new ProtocolValidationError("AgentTurnRow failed protocol validation: missing kind");
  }

  const kind = (value as { kind?: unknown }).kind;
  if (kind === "assistant") {
    return {
      kind,
      ...parseAgentAssistantRow(value),
    };
  }
  if (kind === "toolCall") {
    return {
      kind,
      ...parseAgentToolCallRow(value),
    };
  }
  if (kind === "pendingState") {
    return {
      kind,
      ...parseAgentPendingStateRow(value),
    };
  }

  throw new ProtocolValidationError(
    "AgentTurnRow.kind must be assistant, toolCall, or pendingState",
  );
}

function parseAgentAssistantRow(value: unknown): Omit<AgentAssistantRow, "kind"> {
  if (
    typeof value !== "object" ||
    value === null ||
    !("cursor" in value) ||
    !("sessionId" in value) ||
    !("runId" in value) ||
    !("startedAtMs" in value) ||
    !("completedAtMs" in value) ||
    !("text" in value)
  ) {
    throw new ProtocolValidationError(
      "AgentAssistantRow failed protocol validation: missing required fields",
    );
  }

  const record = value as {
    cursor: unknown;
    sessionId: SessionId;
    runId: string;
    turnId?: string | null;
    startedAtMs: string;
    completedAtMs: string;
    text: string;
  };
  return {
    cursor: parseActivityCursor(record.cursor, "AgentAssistantRow.cursor"),
    sessionId: parseSessionId(record.sessionId),
    runId: parseNonEmptyProtocolString(record.runId, "AgentAssistantRow.runId"),
    turnId: parseNullableBoundaryValue(record.turnId, (turnId) =>
      parseNonEmptyProtocolString(turnId, "AgentAssistantRow.turnId"),
    ),
    startedAtMs: parseProtocolBigInt(record.startedAtMs, "AgentAssistantRow.startedAtMs"),
    completedAtMs: parseProtocolBigInt(record.completedAtMs, "AgentAssistantRow.completedAtMs"),
    text: parseStringField(record.text, "AgentAssistantRow.text"),
  };
}

function parseAgentToolCallRow(value: unknown): Omit<AgentToolCallRow, "kind"> {
  if (
    typeof value !== "object" ||
    value === null ||
    !("cursor" in value) ||
    !("sessionId" in value) ||
    !("runId" in value) ||
    !("toolName" in value) ||
    !("input" in value) ||
    !("output" in value) ||
    !("outcome" in value) ||
    !("startedAtMs" in value) ||
    !("completedAtMs" in value)
  ) {
    throw new ProtocolValidationError(
      "AgentToolCallRow failed protocol validation: missing required fields",
    );
  }

  const record = value as {
    cursor: unknown;
    sessionId: SessionId;
    runId: string;
    turnId?: string | null;
    itemId?: string | null;
    toolName: string;
    input: string;
    output: string;
    outcome: AgentToolCallRow["outcome"];
    startedAtMs: string;
    completedAtMs: string;
  };
  return {
    cursor: parseActivityCursor(record.cursor, "AgentToolCallRow.cursor"),
    sessionId: parseSessionId(record.sessionId),
    runId: parseNonEmptyProtocolString(record.runId, "AgentToolCallRow.runId"),
    turnId: parseNullableBoundaryValue(record.turnId, (turnId) =>
      parseNonEmptyProtocolString(turnId, "AgentToolCallRow.turnId"),
    ),
    itemId: parseNullableBoundaryValue(record.itemId, (itemId) =>
      parseNonEmptyProtocolString(itemId, "AgentToolCallRow.itemId"),
    ),
    toolName: parseStringField(record.toolName, "AgentToolCallRow.toolName"),
    input: parseStringField(record.input, "AgentToolCallRow.input"),
    output: parseStringField(record.output, "AgentToolCallRow.output"),
    outcome: record.outcome,
    startedAtMs: parseProtocolBigInt(record.startedAtMs, "AgentToolCallRow.startedAtMs"),
    completedAtMs: parseProtocolBigInt(record.completedAtMs, "AgentToolCallRow.completedAtMs"),
  };
}

function parseAgentPendingStateRow(value: unknown): Omit<AgentPendingStateRow, "kind"> {
  if (
    typeof value !== "object" ||
    value === null ||
    !("cursor" in value) ||
    !("sessionId" in value) ||
    !("runId" in value) ||
    !("occurredAtMs" in value) ||
    !("state" in value)
  ) {
    throw new ProtocolValidationError(
      "AgentPendingStateRow failed protocol validation: missing required fields",
    );
  }

  const record = value as {
    cursor: unknown;
    sessionId: SessionId;
    runId: string;
    turnId?: string | null;
    occurredAtMs: string;
    state: AgentPendingStateRow["state"];
  };
  return {
    cursor: parseActivityCursor(record.cursor, "AgentPendingStateRow.cursor"),
    sessionId: parseSessionId(record.sessionId),
    runId: parseNonEmptyProtocolString(record.runId, "AgentPendingStateRow.runId"),
    turnId: parseNullableBoundaryValue(record.turnId, (turnId) =>
      parseNonEmptyProtocolString(turnId, "AgentPendingStateRow.turnId"),
    ),
    occurredAtMs: parseProtocolBigInt(record.occurredAtMs, "AgentPendingStateRow.occurredAtMs"),
    state: record.state,
  };
}

File: crates/ta-protocol/src/wire/agent_turns.rs
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use ts_rs::TS;

use crate::wire::{
    ActivityCursor, AgentStreamItemId, AgentStreamTurnId, AgentToolCallOutcome, DaemonEventCursor,
    RunId, RuntimeLanePendingState, SessionId, u64_string,
};

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema, TS)]
#[serde(rename_all = "camelCase")]
#[ts(export_to = "generated/")]
pub struct AgentTurnsPageQuery {
    pub limit: u32,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub before: Option<ActivityCursor>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema, TS)]
#[serde(rename_all = "camelCase")]
#[ts(export_to = "generated/")]
pub struct AgentAssistantRow {
    pub cursor: ActivityCursor,
    pub session_id: SessionId,
    pub run_id: RunId,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub turn_id: Option<AgentStreamTurnId>,
    #[serde(with = "u64_string")]
    #[schemars(schema_with = "u64_string::json_schema")]
    #[ts(as = "u64")]
    pub started_at_ms: u64,
    #[serde(with = "u64_string")]
    #[schemars(schema_with = "u64_string::json_schema")]
    #[ts(as = "u64")]
    pub completed_at_ms: u64,
    pub text: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema, TS)]
#[serde(rename_all = "camelCase")]
#[ts(export_to = "generated/")]
pub struct AgentToolCallRow {
    pub cursor: ActivityCursor,
    pub session_id: SessionId,
    pub run_id: RunId,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub turn_id: Option<AgentStreamTurnId>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub item_id: Option<AgentStreamItemId>,
    pub tool_name: String,
    pub input: String,
    pub output: String,
    pub outcome: AgentToolCallOutcome,
    #[serde(with = "u64_string")]
    #[schemars(schema_with = "u64_string::json_schema")]
    #[ts(as = "u64")]
    pub started_at_ms: u64,
    #[serde(with = "u64_string")]
    #[schemars(schema_with = "u64_string::json_schema")]
    #[ts(as = "u64")]
    pub completed_at_ms: u64,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema, TS)]
#[serde(rename_all = "camelCase")]
#[ts(export_to = "generated/")]
pub struct AgentPendingStateRow {
    pub cursor: ActivityCursor,
    pub session_id: SessionId,
    pub run_id: RunId,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub turn_id: Option<AgentStreamTurnId>,
    #[serde(with = "u64_string")]
    #[schemars(schema_with = "u64_string::json_schema")]
    #[ts(as = "u64")]
    pub occurred_at_ms: u64,
    pub state: RuntimeLanePendingState,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema, TS)]
#[serde(tag = "kind", rename_all = "camelCase")]
#[ts(export_to = "generated/")]
pub enum AgentTurnRow {
    Assistant(AgentAssistantRow),
    ToolCall(AgentToolCallRow),
    PendingState(AgentPendingStateRow),
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema, TS)]
#[serde(rename_all = "camelCase")]
#[ts(export_to = "generated/")]
pub struct AgentTurnsPageResult {
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub items: Vec<AgentTurnRow>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub next_before: Option<ActivityCursor>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub latest_cursor: Option<DaemonEventCursor>,
}
